src/transformers/generation/candidate_generator.py

```diff
@@ -262,6 +348,212 @@ def update_candidate_strategy(self, input_ids: torch.LongTensor, scores: torch.F
+            model as well.
+        inputs_tensor (`torch.Tensor`, *optional*):
+            The model input tensor. In encoder-decoder models, this is the encoder input.
+        assistant_tokenizer (`AutoTokenizer`, *optional*):
+            The tokenizer used for the assistant model.
+        target_tokenizer (`AutoTokenizer`, *optional*):
+            The tokenizer used for the target model.
+        target_lookbehind (int, *optional*):
+            The number of tokens to look behind in the target model.
+        assistant_lookbehind (int, *optional*):
+            The number of tokens to look behind in the assistant model.
+    """
+
+    def __init__(
+        self,
+        input_ids: torch.LongTensor,
+        assistant_model: "PreTrainedModel",
+        generation_config: "GenerationConfig",
+        model_kwargs: Dict,
+        inputs_tensor: Optional[torch.Tensor] = None,
+        logits_processor: "LogitsProcessorList" = None,
+        assistant_tokenizer: Optional[AutoTokenizer] = None,
+        target_tokenizer: Optional[AutoTokenizer] = None,
+        target_lookbehind: Optional[int] = 20,
+        assistant_lookbehind: Optional[int] = 2,
+    ):
+        super().__init__(input_ids, assistant_model, generation_config, model_kwargs, inputs_tensor, logits_processor)
+
+        self.assistant_tokenizer = assistant_tokenizer
+        self.target_tokenizer = target_tokenizer
+        self.target_lookbehind = target_lookbehind
+        self.assistant_lookbehind = assistant_lookbehind
+
+        self.prev_tokens = None
+
+    def convert_token_ids(
+        self,
+        input_ids,
+        src,
+        dest,
+    ):
+        """
+        Convert token IDs from one tokenizer to another.
+        Args:
+            input_ids: The input token IDs.
+            src: The source tokenizer.
+            dest: The destination tokenizer.
+        Returns:
+            The converted token IDs.
+        """
+        text = src.batch_decode(input_ids, skip_special_tokens=True, clean_up_tokenization_spaces=True)
+        dest_ids = dest(text, add_special_tokens=True, return_tensors="pt")["input_ids"]
+        return dest_ids.to(input_ids.device)
+
+    def get_candidates(
+        self, input_ids: torch.LongTensor, stopping_criteria
+    ) -> Tuple[torch.LongTensor, Optional[torch.FloatTensor]]:
+        input_ids = input_ids.to(self.assistant_model.device)
+        convert_kwargs = {"src": self.target_tokenizer, "dest": self.assistant_tokenizer}
+        remove_from_pkv = 0
+
+        if self.prev_tokens is None:
+            assistant_input_ids = self.convert_token_ids(input_ids, **convert_kwargs)
+            self.prev_target_ids = input_ids
+            self.prev_assistant_ids = assistant_input_ids
+            new_cur_len = assistant_input_ids.shape[-1]
+        else:
+            # input_ids contains all target prompt input ids and some new target input ids
+            if self.prev_target_ids.shape[1] > self.target_lookbehind:
```

Comment:
Sanity check: you have confirmed that the runtime with this logic is faster than simply re-encoding the whole `input_ids`?

If it is just a tiny bit faster, like 5%, then I'd prefer having the simpler reencoding code path exclusively -- it's much simpler to maintain and document

(I'll likely have a few comments to make this section more beginner-friendly, but I'm holding them until we confirm this part :D)

Reply:
Yes, we have tested both options and this one is significantly faster.
Natuarlly, as the number of new tokens increases it becomes even more significant.

Reply:
In one test we did, it was 15% faster. And that test used `cnn_dailymail` which does not even have long prompts.

Re-encoding (decode+encode) the whole `input_ids` (which increase in size at at each step) at every assistant generation step incurs a significant overhead.

Reply:
awesome, thank you for confirming

(going to leave this comment unresolved for future reference)